engine/src/main/java/org/kigalisim/engine/state/StreamKeeper.java

```diff
@@ -0,0 +1,652 @@
+        manufactureAmount.divide(totalAmount);
+    BigDecimal importPercent = isZero ? new BigDecimal("0.5") : 
+        importAmount.divide(totalAmount);
+    
+    BigDecimal manufactureShare = amountKg.multiply(manufacturePercent);
+    BigDecimal importShare = amountKg.multiply(importPercent);
+    EngineNumber manufactureNewValue = new EngineNumber(manufactureShare, value.getUnits());
+    EngineNumber importNewValue = new EngineNumber(importShare, value.getUnits());
+    
+    setStream(application, substance, "manufacture", manufactureNewValue);
+    setStream(application, substance, "import", importNewValue);
+  }
+  
+  /**
+   * Determine if the user is setting a sales component (manufacture / import / sales) by units.
+   *
+   * @param name The stream name
+   * @param value The value to set
+   * @return true if the user is setting a sales component by units and false otherwise
+   */
+  private boolean getIsSettingVolumeByUnits(String name, EngineNumber value) {
+    boolean isSalesComponent = "manufacture".equals(name) || "import".equals(name) 
+                               || "sales".equals(name);
+    boolean isUnits = value.getUnits().startsWith("unit");
+    return isSalesComponent && isUnits;
+  }
+  
+  /**
+   * Handle setting a stream which only requires simple unit conversion.
+   *
+   * @param application The application name
+   * @param substance The substance name
+   * @param name The stream name
+   * @param value The value to set
+   */
+  private void setSimpleStream(String application, String substance, String name, 
+                              EngineNumber value) {
+    String unitsNeeded = getUnits(name);
+    EngineNumber valueConverted = unitConverter.convert(value, unitsNeeded);
+    
+    if (CHECK_NAN_STATE && valueConverted.getValue().toString().equals("NaN")) {
+      String pieces = String.join(" > ", application, substance, name);
+      throw new RuntimeException("Encountered NaN after conversion to be set for: " + pieces);
+    }
+    
+    if (CHECK_POSITIVE_STREAMS && valueConverted.getValue().compareTo(BigDecimal.ZERO) < 0) {
+      String pieces = String.join(" > ", application, substance, name);
+      throw new RuntimeException("Encountered negative stream to be set for: " + pieces);
+    }
+    
+    streams.put(getKey(application, substance, name), valueConverted);
+  }
+  
+  /**
+   * Handle setting volume by units for sales components.
+   *
+   * <p>Handle setting a sales component (manufacture or import) which requires conversion
+   * by way of initial charge specific to that stream.</p>
+   *
+   * @param application The application name
+   * @param substance The substance name
+   * @param name The stream name
+   * @param value The value to set
+   */
+  private void setStreamForSalesWithUnits(String application, String substance, String name, 
+                                         EngineNumber value) {
+    // For now, this is a simplified implementation
```

Comment:
Please reply further explaining this comment and if another follow up PR is needed so that I can create a task. What are we waiting on to fully complete this?